### ISSUE TITLE
v0.2.0_1 is 10% slower than v0.1.12_2, why?

### ISSUE THREAD

**Issue by xuancong84** [pytorch/pytorch#2338]
I have benchmark tested v0.2.0_1 vs v0.1.12_2 using OpenNMT, https://github.com/OpenNMT/OpenNMT-py

For training, on exactly the same data, same model size, same total number of parameters, etc., v0.2.0_1 uses 10% more time than v0.1.12_2.

I hope you guys be aware of performance drop when implementing new features.

Any ideas, thanks!

**Comment by fmassa** [comment]
This might be due to broadcasting, see https://github.com/pytorch/pytorch/issues/2012

**Comment by gchanan** [comment]
Thanks for the benchmark; I'll compare it when looking at https://github.com/pytorch/pytorch/issues/2012.

**Comment by xuancong84** [comment]
I have just finished benchmarking the correctness/accuracy. Fortunately, the accuracy is the same.
For OpenNMT-py, on 1M parallel sentences, using 50k num_ops for subword:
**v0.1.12_2:**
Epoch 15, 12250/12296; acc:  49.58; ppl:  13.18; 6873 src tok/s; 4762 tgt tok/s;  66690 s elapsed
Train perplexity: 12.9717                                                                        
Train accuracy: 49.8811                                                                          
Validation perplexity: 22.4254                                                                   
Validation accuracy: 46.0955      
For testing (with beam search)
BLEU 0.1176               
NIST 4.4079               
TER 0.748756287277025     
METEOR 0.20517773332505518                                                 

**v0.2.0_1:**
Epoch 15, 12250/12296; acc:  50.47; ppl:  12.59; 6505 src tok/s; 4514 tgt tok/s;  73246 s elapsed
Train perplexity: 12.9133                                                                        
Train accuracy: 49.9525                                                                          
Validation perplexity: 22.5645                                                                   
Validation accuracy: 45.7346                                                                     
For testing (with beam search)
BLEU 0.1144               
NIST 4.2997               
TER 0.7462276337850094    
METEOR 0.20318649362871716

The difference is within the random fluctuation margin. So I can conclude that the new version can achieve the same accuracy in calculations.

**Comment by xuancong84** [comment]
Sorry. I just found out that it is because the two computing nodes run at different speeds even though they have identical hardware and OS. The newer version, v0.2.0_1, is in fact about 3% faster than the old version v0.1.12_2 .
Not sure what you guys have found out.
